File: laravel-api/app/Models/Message.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasOne;
use Illuminate\Database\Eloquent\Relations\MorphMany;
use Illuminate\Database\Eloquent\SoftDeletes;
use Illuminate\Database\Eloquent\Builder;

class Message extends Model
{
    use SoftDeletes;

    // Message types constants
    const TYPE_TEXT = 'text';
    const TYPE_IMAGE = 'image';
    const TYPE_FILE = 'file';
    const TYPE_VIDEO = 'video';
    const TYPE_AUDIO = 'audio';
    const TYPE_SYSTEM = 'system';

    // Message status constants
    const STATUS_SENT = 'sent';
    const STATUS_DELIVERED = 'delivered';
    const STATUS_READ = 'read';
    const STATUS_FAILED = 'failed';

    protected $fillable = [
        'conversation_id',
        'sender_id',
        'reply_to_message_id',
        'message',
        'type',
        'status',
        'file_path',
        'file_name',
        'file_type',
        'file_size',
        'is_read',
        'is_edited',
        'edited_at',
        'delivered_at',
        'read_at',
        'is_pinned',
    ];

    protected $casts = [
        'is_read' => 'boolean',
        'is_edited' => 'boolean',
        'edited_at' => 'datetime',
        'delivered_at' => 'datetime',
        'read_at' => 'datetime',
        'is_pinned' => 'boolean',
        'created_at' => 'datetime',
        'updated_at' => 'datetime',
        'deleted_at' => 'datetime',
    ];

    protected $appends = [
        'is_file_message',
        'formatted_file_size',
    ];

    /**
     * Relationships
     */

    public function conversation(): BelongsTo
    {
        return $this->belongsTo(Conversation::class);
    }

    public function sender(): BelongsTo
    {
        return $this->belongsTo(WpUser::class, 'sender_id', 'ID');
    }

    public function replyTo(): BelongsTo
    {
        return $this->belongsTo(Message::class, 'reply_to_message_id');
    }

    public function replies()
    {
        return $this->hasMany(Message::class, 'reply_to_message_id');
    }

    /**
     * Get all of the message's reactions.
     */
    public function reactions(): MorphMany
    {
        return $this->morphMany(MessageReaction::class, 'reactable');
    }

    /**
     * Scopes
     */

    public function scopeUnread(Builder $query): Builder
    {
        return $query->where('is_read', false);
    }

    public function scopeRead(Builder $query): Builder
    {
        return $query->where('is_read', true);
    }

    public function scopeByType(Builder $query, string $type): Builder
    {
        return $query->where('type', $type);
    }

    public function scopeTextMessages(Builder $query): Builder
    {
        return $query->where('type', self::TYPE_TEXT);
    }

    public function scopeFileMessages(Builder $query): Builder
    {
        return $query->whereIn('type', [
            self::TYPE_IMAGE,
            self::TYPE_FILE,
            self::TYPE_VIDEO,
            self::TYPE_AUDIO
        ]);
    }

    public function scopeForConversation(Builder $query, int $conversationId): Builder
    {
        return $query->where('conversation_id', $conversationId);
    }

    public function scopeFromSender(Builder $query, int $senderId): Builder
    {
        return $query->where('sender_id', $senderId);
    }

    public function scopeLatestFirst(Builder $query): Builder
    {
        return $query->orderBy('created_at', 'desc');
    }

    public function scopeOldestFirst(Builder $query): Builder
    {
        return $query->orderBy('created_at', 'asc');
    }

    public function scopeSearchByContent(Builder $query, string $search): Builder
    {
        return $query->where('message', 'like', "%{$search}%");
    }

    /**
     * Accessors
     */

    public function getIsFileMessageAttribute(): bool
    {
        return in_array($this->type, [
            self::TYPE_IMAGE,
            self::TYPE_FILE,
            self::TYPE_VIDEO,
            self::TYPE_AUDIO
        ]);
    }

    public function getFormattedFileSizeAttribute(): ?string
    {
        if (!$this->file_size) {
            return null;
        }

        $units = ['B', 'KB', 'MB', 'GB'];
        $size = $this->file_size;
        $unitIndex = 0;

        while ($size >= 1024 && $unitIndex < count($units) - 1) {
            $size /= 1024;
            $unitIndex++;
        }

        return round($size, 2) . ' ' . $units[$unitIndex];
    }

    /**
     * Helper Methods
     */

    public function markAsRead(): bool
    {
        if ($this->is_read) {
            return true;
        }

        return $this->update([
            'is_read' => true,
            'status' => self::STATUS_READ,
            'read_at' => now(),
        ]);
    }

    public function markAsDelivered(): bool
    {
        if ($this->status === self::STATUS_DELIVERED || $this->status === self::STATUS_READ) {
            return true;
        }

        return $this->update([
            'status' => self::STATUS_DELIVERED,
            'delivered_at' => now(),
        ]);
    }

    public function markAsFailed(): bool
    {
        return $this->update([
            'status' => self::STATUS_FAILED,
        ]);
    }

    public function isTextMessage(): bool
    {
        return $this->type === self::TYPE_TEXT;
    }

    public function isImageMessage(): bool
    {
        return $this->type === self::TYPE_IMAGE;
    }

    public function isFileMessage(): bool
    {
        return $this->type === self::TYPE_FILE;
    }

    public function isVideoMessage(): bool
    {
        return $this->type === self::TYPE_VIDEO;
    }

    public function isAudioMessage(): bool
    {
        return $this->type === self::TYPE_AUDIO;
    }

    public function isSystemMessage(): bool
    {
        return $this->type === self::TYPE_SYSTEM;
    }

    public function hasFile(): bool
    {
        return !empty($this->file_path);
    }

    public function isEdited(): bool
    {
        return $this->is_edited;
    }

    public function isReply(): bool
    {
        return !empty($this->reply_to_message_id);
    }

    public function getFileUrl(): ?string
    {
        if (!$this->file_path) {
            return null;
        }

        return asset('storage/' . $this->file_path);
    }

    /**
     * Update message content and mark as edited
     */
    public function updateContent(string $newMessage): bool
    {
        return $this->update([
            'message' => $newMessage,
            'is_edited' => true,
            'edited_at' => now(),
        ]);
    }

    /**
     * Check if message contains embedded images
     */
    public function hasEmbeddedImages(): bool
    {
        return preg_match('/\[IMAGE\].*?\[\/IMAGE\]/s', $this->message) === 1;
    }

    /**
     * Get embedded image URLs from message
     */
    public function getEmbeddedImages(): array
    {
        $images = [];
        preg_match_all('/\[IMAGE\](.*?)\[\/IMAGE\]/s', $this->message, $matches);

        if (!empty($matches[1])) {
            $images = $matches[1];
        }

        return $images;
    }

    /**
     * Get message text without image markup
     */
    public function getTextContent(): string
    {
        return trim(preg_replace('/\[IMAGE\].*?\[\/IMAGE\]/s', '', $this->message));
    }

    /**
     * Get parsed message content with text and images separated
     */
    public function getParsedContent(): array
    {
        return [
            'text' => $this->getTextContent(),
            'images' => $this->getEmbeddedImages(),
            'has_images' => $this->hasEmbeddedImages(),
        ];
    }
}

File: flutter/lib/src/models/message.dart
class Message {
  final int id;
  final int conversationId;
  final int senderId;
  final int? replyToMessageId;
  final String message;
  final String type;
  final String status;
  final String? filePath;
  final String? fileName;
  final String? fileType;
  final int? fileSize;
  final bool isRead;
  final bool isEdited;
  final bool isPinned;
  final String? editedAt;
  final String? deliveredAt;
  final String? readAt;
  final String? createdAt;
  final String? updatedAt;

  // Sender info (matching Laravel API response)
  final Map<String, dynamic>? sender;

  // Whether this message is from the current user
  final bool isMine;

  // Reply to message (if loaded)
  final Map<String, dynamic>? replyTo;

  // Reactions
  final int reactionsCount;
  final List<String>? reactions;
  final String? myReaction;


  // Message types
  static const String typeText = 'text';
  static const String typeImage = 'image';
  static const String typeFile = 'file';
  static const String typeVideo = 'video';
  static const String typeAudio = 'audio';
  static const String typeSystem = 'system';

  // Message statuses
  static const String statusSent = 'sent';
  static const String statusDelivered = 'delivered';
  static const String statusRead = 'read';
  static const String statusFailed = 'failed';

  Message({
    required this.id,
    required this.conversationId,
    required this.senderId,
    this.replyToMessageId,
    required this.message,
    this.type = typeText,
    this.status = statusSent,
    this.filePath,
    this.fileName,
    this.fileType,
    this.fileSize,
    this.isRead = false,
    this.isEdited = false,
    this.isPinned = false,
    this.editedAt,
    this.deliveredAt,
    this.readAt,
    this.createdAt,
    this.updatedAt,
    this.sender,
    this.isMine = false,
    this.replyTo,
    this.reactionsCount = 0,
    this.reactions,
    this.myReaction,
  });

  factory Message.fromJson(Map<String, dynamic> json) {
    try {
      // Helper function to parse int from string or int
      int? _parseInt(dynamic value) {
        if (value == null) return null;
        if (value is int) return value;
        if (value is String) return int.tryParse(value);
        return null;
      }

      // Get senderId from either sender_id field or sender object
      int senderId = 0;
      if (json['sender_id'] != null) {
        senderId = _parseInt(json['sender_id']) ?? 0;
      } else if (json['sender'] != null && json['sender'] is Map) {
        final senderMap = json['sender'] as Map<String, dynamic>;
        if (senderMap['id'] != null) {
          senderId = _parseInt(senderMap['id']) ?? 0;
        }
      }

      // Safe parse sender map
      Map<String, dynamic>? senderMap;
      if (json['sender'] != null && json['sender'] is Map) {
        senderMap = Map<String, dynamic>.from(json['sender']);
      }

      // Safe parse replyTo map
      Map<String, dynamic>? replyToMap;
      if (json['reply_to'] != null && json['reply_to'] is Map) {
        replyToMap = Map<String, dynamic>.from(json['reply_to']);
      }

      // Safe parse reactions list
      List<String>? reactionsList;
      if (json['reactions'] != null && json['reactions'] is List) {
        reactionsList = (json['reactions'] as List).map((e) => e.toString()).toList();
      }

      return Message(
        id: _parseInt(json['id']) ?? 0,
        conversationId: _parseInt(json['conversation_id']) ?? 0,
        senderId: senderId,
        replyToMessageId: _parseInt(json['reply_to_message_id']),
        message: json['message'] as String? ?? '',
        type: json['type'] as String? ?? typeText,
        status: json['status'] as String? ?? statusSent,
        filePath: json['file_path'] as String?,
        fileName: json['file_name'] as String?,
        fileType: json['file_type'] as String?,
        fileSize: _parseInt(json['file_size']),
        isRead: json['is_read'] == true || json['is_read'] == 1 || json['is_read'] == '1',
        isEdited: json['is_edited'] == true || json['is_edited'] == 1 || json['is_edited'] == '1',
        isPinned: json['is_pinned'] == true || json['is_pinned'] == 1 || json['is_pinned'] == '1',
        editedAt: json['edited_at'] as String?,
        deliveredAt: json['delivered_at'] as String?,
        readAt: json['read_at'] as String?,
        createdAt: json['created_at'] as String?,
        updatedAt: json['updated_at'] as String?,
        sender: senderMap,
        isMine: json['is_mine'] == true || json['is_mine'] == 1 || json['is_mine'] == '1',
        replyTo: replyToMap,
        reactionsCount: _parseInt(json['reactions_count']) ?? 0,
        reactions: reactionsList,
        myReaction: json['my_reaction'] as String?,
      );
    } catch (e, stackTrace) {
      print('Message.fromJson - Error parsing: $e');
      print('Message.fromJson - JSON data: $json');
      print('Message.fromJson - Stack trace: $stackTrace');
      rethrow;
    }
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'conversation_id': conversationId,
      'sender_id': senderId,
      'reply_to_message_id': replyToMessageId,
      'message': message,
      'type': type,
      'status': status,
      'file_path': filePath,
      'file_name': fileName,
      'file_type': fileType,
      'file_size': fileSize,
      'is_read': isRead,
      'is_edited': isEdited,
      'is_pinned': isPinned,
      'edited_at': editedAt,
      'delivered_at': deliveredAt,
      'read_at': readAt,
      'created_at': createdAt,
      'updated_at': updatedAt,
      if (sender != null) 'sender': sender,
      'is_mine': isMine,
      if (replyTo != null) 'reply_to': replyTo,
      'reactions_count': reactionsCount,
      if (reactions != null) 'reactions': reactions,
      if (myReaction != null) 'my_reaction': myReaction,
    };
  }

  bool isTextMessage() => type == typeText;
  bool isImageMessage() => type == typeImage;
  bool isFileMessage() => type == typeFile;
  bool isVideoMessage() => type == typeVideo;
  bool isAudioMessage() => type == typeAudio;
  bool isSystemMessage() => type == typeSystem;

  bool hasFile() => filePath != null && filePath!.isNotEmpty;
  bool isReply() => replyToMessageId != null;

  String? get formattedFileSize {
    if (fileSize == null) return null;

    const units = ['B', 'KB', 'MB', 'GB'];
    double size = fileSize!.toDouble();
    int unitIndex = 0;

    while (size >= 1024 && unitIndex < units.length - 1) {
      size /= 1024;
      unitIndex++;
    }

    return '${size.toStringAsFixed(2)} ${units[unitIndex]}';
  }

  Message copyWith({
    int? id,
    int? conversationId,
    int? senderId,
    int? replyToMessageId,
    String? message,
    String? type,
    String? status,
    String? filePath,
    String? fileName,
    String? fileType,
    int? fileSize,
    bool? isRead,
    bool? isEdited,
    bool? isPinned,
    String? editedAt,
    String? deliveredAt,
    String? readAt,
    String? createdAt,
    String? updatedAt,
    Map<String, dynamic>? sender,
    bool? isMine,
    Map<String, dynamic>? replyTo,
    int? reactionsCount,
    List<String>? reactions,
    String? myReaction,
  }) {
    return Message(
      id: id ?? this.id,
      conversationId: conversationId ?? this.conversationId,
      senderId: senderId ?? this.senderId,
      replyToMessageId: replyToMessageId ?? this.replyToMessageId,
      message: message ?? this.message,
      type: type ?? this.type,
      status: status ?? this.status,
      filePath: filePath ?? this.filePath,
      fileName: fileName ?? this.fileName,
      fileType: fileType ?? this.fileType,
      fileSize: fileSize ?? this.fileSize,
      isRead: isRead ?? this.isRead,
      isEdited: isEdited ?? this.isEdited,
      isPinned: isPinned ?? this.isPinned,
      editedAt: editedAt ?? this.editedAt,
      deliveredAt: deliveredAt ?? this.deliveredAt,
      readAt: readAt ?? this.readAt,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      sender: sender ?? this.sender,
      isMine: isMine ?? this.isMine,
      replyTo: replyTo ?? this.replyTo,
      reactionsCount: reactionsCount ?? this.reactionsCount,
      reactions: reactions ?? this.reactions,
      myReaction: myReaction ?? this.myReaction,
    );
  }

  // Convenience getters for sender info
  String? get senderName => sender?['name'] as String?;
  int? get senderIdFromMap => sender?['id'] as int?;
}
